### ISSUE TITLE
Incorrect IDE0059 when local used in lambda before reassignment

### ISSUE THREAD

**Issue by YairHalberstadt** [dotnet/roslyn#42334]
**Version Used**: master as of the 10th march

**Steps to Reproduce**:

```
using System;
public static class Program {
    public static void Main() {
        var x = default(int);
        Func<string> a = () => x.ToString();
        x = 41;
        Console.WriteLine(a());
    }
}
```

**Expected Behavior**: No warning.

**Actual Behavior**: IDE0059 unnecessary assignment of a value to 'x'
Suggested codefix 'remove redundant assignment' doesn't compile:

```csharp
using System;
public static class Program {
    public static void Main() {
        int x;
        Func<string> a = () => x.ToString();
        x = 41;
        Console.WriteLine(a());
    }
}
```

**Comment by mavasani** [comment]
I believe this is by design - the analyzer is actually performing dataflow analysis here and knows that the first invocation to the lambda does not happen till after the second assignment. Basically, there is no reads for `x` in between the two writes, so the first write is indeed redundant.

**Comment by mavasani** [comment]
Ah, I didn't notice that compiler flags an error `CS0165` here, which seems to be purely based on lexical analysis. We will have to tweak the analyzer here to bail out for this case.

**Comment by mavasani** [comment]
Acutally, putting some more thought here, I believe the analyzer diagnostic here can indeed by helpful. Assume the first assignment to `x` required some expensive computation. Given that at runtime that value is indeed never read, the above code can be rewritten to as follows to avoid redundant computation:

Original code:
```csharp
using System;
public static class Program
{
    public static void Main()
    {
        var x = SuperExpensiveComputation();
        Func<string> a = () => x.ToString();
        x = 41;
        Console.WriteLine(a());
    }

    private static int SuperExpensiveComputation() => 0;
}
```

Refactored code:
1. If `SuperExpensiveComputation` has no side effects:
```csharp
using System;
public static class Program
{
    public static void Main()
    {
        int x;
        string a() => x.ToString();
        x = 41;
        Console.WriteLine(a());
    }

    private static int SuperExpensiveComputation() => 0;
}
```
2. If `SuperExpensiveComputation` has side effects:
```csharp
using System;
public static class Program
{
    public static void Main()
    {
        _ = SuperExpensiveComputation();
        int x;
        string a() => x.ToString();
        x = 41;
        Console.WriteLine(a());
    }

    private static int SuperExpensiveComputation() => 0;
}
```

**Comment by YairHalberstadt** [comment]
@mavasani 

The refactored code does not compile though

**Comment by YairHalberstadt** [comment]
Also there is a case which is definitely a bug:

```csharp
using System;
public static class Program {
    public static void Main() {
        var x = 42;
        Func<string> a = x.ToStringExt;
        x = 10;
        Console.WriteLine(a());
    }
    
    public static string ToStringExt(this object x) => x.ToString();
}
```

Here `42` is printed, not `10`, but the compiler still reports IDE0059

**Comment by mavasani** [comment]
@YairHalberstadt Refactored code compiles fine for me: 
![LocalFunction](https://user-images.githubusercontent.com/10605811/76425758-50681080-6367-11ea-89c7-bc31bbcd169d.png)

The repro case you provided in https://github.com/dotnet/roslyn/issues/42334#issuecomment-597653971 is indeed a bug in the analyzer, where the initialization is indeed not redundant. Can you please file a separate issue for it?

**Comment by YairHalberstadt** [comment]
Sorry, I was mistaken about not compiling, and will file a separate issue.
Thanks for your input!

**Comment by YairHalberstadt** [comment]
https://github.com/dotnet/roslyn/issues/42337
